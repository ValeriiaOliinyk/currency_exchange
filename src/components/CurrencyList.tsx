import React from "react";
import PropTypes from "prop-types";
import { useSelector, useDispatch } from "react-redux";

import {
  addFavorite,
  deleteFavorite,
  getFavorites,
  getRegularCurrency,
  getNumberOfFavorites,
} from "../redux/ducks/currency";

// Components
import { MainLoader, Empty } from ".";
import {
  ListFavorite,
  BtnFavorite,
  BtnEmpty,
  ItemFavorites,
  ItemContainer,
  Value,
} from "../styles/components";

export function CurrencyList() {
  const dispatch = useDispatch();
  const favorites: Array<string> | undefined = useSelector(getFavorites);
  const regularCurrency: Array<string> | undefined = useSelector(
    getRegularCurrency
  );
  const numberOfFavorites: Number | undefined = useSelector(
    getNumberOfFavorites
  );

  const toggleFavAction = (text: string) => {
    if (!favorites.includes(text)) {
      dispatch(addFavorite(text));
    }

    if (favorites.includes(text)) {
      dispatch(deleteFavorite(text));
    }
  };

  return (
    <>
      <ListFavorite favorite>
        {numberOfFavorites > 0 ? (
          favorites.map((item) => (
            <ItemFavorites key={item}>
              <ItemContainer>
                <Value>{item}</Value>
                <BtnFavorite
                  className="Btn"
                  type="button"
                  onClick={() => toggleFavAction(item)}
                ></BtnFavorite>
              </ItemContainer>
            </ItemFavorites>
          ))
        ) : (
          <Empty />
        )}
      </ListFavorite>
      <ListFavorite>
        {regularCurrency ? (
          regularCurrency.map((item) => (
            <ItemFavorites key={item}>
              <ItemContainer>
                <Value>{item}</Value>
                <BtnEmpty
                  type="button"
                  onClick={() => toggleFavAction(item)}
                ></BtnEmpty>
              </ItemContainer>
            </ItemFavorites>
          ))
        ) : (
          <MainLoader />
        )}
      </ListFavorite>
    </>
  );
}

CurrencyList.defaultProps = {
  numberOfFavorites: 0,
  favorites: [],
};

CurrencyList.propTypes = {
  numberOfFavorites: PropTypes.number,
  favorites: PropTypes.arrayOf(PropTypes.string),
  regularCurrency: PropTypes.arrayOf(PropTypes.string),
};
