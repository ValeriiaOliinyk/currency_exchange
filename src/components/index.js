export { Convert } from "./Convert/Convert";
export { CurrencyList } from "./CurrencyList/CurrencyList";
export { CurrencyRow } from "./CurrencyRow/CurrencyRow";
export { Empty } from "./Empty/Empty";
export { MainLoader } from "./MainLoader/MainLoader";
export { Navigation } from "./Navigation/Navigation";
