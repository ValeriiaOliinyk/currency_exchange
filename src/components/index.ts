export { Convert } from "./Convert";
export { CurrencyList } from "./CurrencyList";
export { CurrencyRow } from "./CurrencyRow";
export { Empty } from "./Empty";
export { MainLoader } from "./MainLoader";
export { Navigation } from "./Navigation";
export { Gallery } from "./Gallery";
export { Chart } from "./Chart";
