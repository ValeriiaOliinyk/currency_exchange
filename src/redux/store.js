import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from "@reduxjs/toolkit";
import logger from "redux-logger";
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from "redux-persist";
import storage from "redux-persist/lib/storage";
import createSagaMiddleware from "redux-saga";
import { exchangeRateReducer } from "./currency/currency-reducers";
import { updateDataReducer } from "./currency/dataUrl-reducer";
import { favoriteReducer } from "./currency/favorite-reducer";
import { currencyReducer } from "./currency/fetch-reducer";
import { sagaWatcher } from "./currency/fetch-reducer";

const saga = createSagaMiddleware();

const persistConfig = {
  key: "favoritesList",
  storage,
  blacklist: ["currency", "data", "rate"],
};

const middleware = [
  ...getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  logger,
  saga,
];

const rootReducer = combineReducers({
  currency: currencyReducer,
  favorite: favoriteReducer,
  data: updateDataReducer,
  rate: exchangeRateReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware,
  devTools: process.env.NODE_ENV === "development",
});

saga.run(sagaWatcher);

const persistor = persistStore(store);

export default { store, persistor };
