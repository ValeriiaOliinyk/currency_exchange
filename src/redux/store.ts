import {
  combineReducers,
  configureStore,
  getDefaultMiddleware,
} from "@reduxjs/toolkit";
import logger from "redux-logger";
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from "redux-persist";
import storage from "redux-persist/lib/storage";
import createSagaMiddleware from "redux-saga";
import {
  exchangeRateReducer,
  updateDataReducer,
  favoriteReducer,
  currencyReducer,
  sagaWatcher,
  putUpdatedDataReducer,
  watchUpdateData,
} from "./ducks/currency";
const saga = createSagaMiddleware();

const persistConfig = {
  key: "favoritesList",
  storage,
  blacklist: ["currency", "data", "updatedData"],
};

const middleware = [
  ...getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
  logger,
  saga,
];

const rootReducer = combineReducers({
  currency: currencyReducer,
  favorite: favoriteReducer,
  data: updateDataReducer,
  rate: exchangeRateReducer,
  updatedData: putUpdatedDataReducer,
});

type RootReducerType = typeof rootReducer;

export type AppStateType = ReturnType<RootReducerType>;

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  reducer: persistedReducer,
  middleware,
  devTools: process.env.NODE_ENV === "development",
});

saga.run(sagaWatcher);
saga.run(watchUpdateData);

const persistor = persistStore(store);

export default { store, persistor };
