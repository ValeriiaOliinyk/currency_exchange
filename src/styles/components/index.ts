export { BtnEmpty } from './BtnEmpty';
export { BtnFavorite } from './BtnFavorite';
export { Container } from './Container';
export { EmptyFavorites } from './EmptyFavorites';
export { Favorite } from './Favorite';
export { FormBox } from './FormBox';
export { SelectField } from './SelectField';
export { FormControlComponent } from './FormControlComponent';
export { FormCurrency } from './FormCurrency';
export { FormEqually } from './FormEqually';
export { Header } from './Header';
export { ItemContainer } from './ItemContainer';
export { ItemFavorites } from './ItemFavorites';
export { ListFavorite } from './ListFavorite';
export { LoaderMain } from './LoaderMain';
export { Logo } from './Logo';
export { Main } from './Main';
export { MainTitle } from './MainTitle';
export { Nav } from './Nav';
export { NavLinks } from './NavLink';
export { NavigationItems } from './NavigationItems';
export { NavigationList } from './NavigationList';
export { Option } from './Option';
export { Title } from './Title';
export { TitleHome } from './TitleHome';
export { Value } from './Value';
export { ErrorMessage } from './ErrorMessage';
export { FormControlBox } from './FormControlBox';
export { Image } from './Image';
