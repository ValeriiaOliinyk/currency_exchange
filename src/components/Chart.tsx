import React from "react";
import { Bar } from "react-chartjs-2";

import { chart } from "../helpers/chartData";
import { Container, Title } from "../styles/components";

export const Chart = () => {
  return (
    <>
      <Container>
        <Title>The top 10 most traded currencies in the world</Title>
        <Bar
          data={chart.data}
          width={50}
          height={100}
          options={{
            legend: {
              display: true,
              position: "bottom"
            }
          }}
        />
      </Container>
    </>
  );
};
