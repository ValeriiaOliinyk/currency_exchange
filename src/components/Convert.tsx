import React from 'react';

// Components
import { CurrencyRow } from './CurrencyRow';
import { Title, Container } from '../styles/components';

export function Convert() {
  return (
    <Container>
      <Title>Convert</Title>
      <CurrencyRow />
    </Container>
  );
}
