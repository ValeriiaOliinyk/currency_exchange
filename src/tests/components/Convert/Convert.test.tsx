import React from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';
import { render, screen } from '@testing-library/react';

import store from '../../../redux/store';

import { Convert } from '../../../components';
import Theme from '../../../styles/theme';

it('renders with text', () => {
  render(
    <ThemeProvider theme={Theme}>
      <Provider store={store.store}>
        <Convert />
      </Provider>
    </ThemeProvider>,
  );
  screen.debug();
  expect(screen.getByText('Convert')).toBeInTheDocument();
});
