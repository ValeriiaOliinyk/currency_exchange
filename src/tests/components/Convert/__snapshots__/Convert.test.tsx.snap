// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container component works 1`] = `
.c0 {
  width: 80%;
  margin: auto;
  margin-top: 100px;
}

<section
  className="c0"
/>
`;

exports[`CurrencyRow component works 1`] = `
<ContextProvider
  value={
    Object {
      "breakpoints": Object {
        "mobileL": "screen and (max-width: 425px)",
        "tablet": "screen and (max-width: 768px)",
      },
      "colors": Object {
        "black": "#000000",
        "dark": "#56809b",
        "hover": "#67a030",
        "main": "#a6dbfb",
        "red": "#800000",
        "text": "#0e2b3c",
        "transparent": "#ffffff00",
        "white": "#ffffff",
      },
    }
  }
>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Convert />
  </Provider>
</ContextProvider>
`;

exports[`First test with react testing library 1`] = `
<DocumentFragment>
  .c0 {
  width: 80%;
  margin: auto;
  margin-top: 100px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 {
  outline: none;
  width: 75%;
  padding: 10px;
  border-radius: 6px;
  border: 1px solid #000000;
  border-color: #000000;
}

.c2 {
  margin: auto;
  width: 50%;
  height: 300px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
  padding: 20px;
  border-radius: 20px;
}

.c7 {
  width: 40px;
  height: 40px;
  background-image: url("./images/equal.svg");
  text-align: center;
  font-size: 30px;
  margin-bottom: 26px;
}

.c5 {
  width: 20%;
}

.c1 {
  text-align: center;
  margin-bottom: 30px;
  color: #67a030;
  font-weight: 700;
  font-size: 32px;
}

.c6 {
  width: 100%;
  font-size: 18px;
  font-weight: 700;
  color: #ffffff00;
}

@media screen and (max-width:768px) {
  .c2 {
    width: 90%;
  }
}

@media screen and (max-width:425px) {
  .c5 {
    width: 40%;
  }
}

@media screen and (max-width:768px) {
  .c5 {
    font-size: 12px;
  }
}

@media screen and (max-width:768px) {
  .c1 {
    font-size: 22px;
  }
}

<section
    class="c0"
  >
    <h2
      class="c1"
    >
      Convert
    </h2>
    <div
      class="container"
    >
      <form
        class="c2"
      >
        <div
          class="c3"
        >
          <input
            class="c4"
            name="fromAmount"
            type="number"
            value="1"
          />
          <select
            class="c5"
          />
        </div>
        <div
          class="c6"
        >
          Transparent
        </div>
        <div
          class="c7"
        />
        <div
          class="c3"
        >
          <input
            class="c4"
            name="toAmount"
            type="number"
            value="0.0"
          />
          <select
            class="c5"
          />
        </div>
        <div
          class="c6"
        >
          Transparent
        </div>
      </form>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Renders correctly 1`] = `
.c0 {
  text-align: center;
  margin-bottom: 30px;
  color: #67a030;
  font-weight: 700;
  font-size: 32px;
}

@media screen and (max-width:768px) {
  .c0 {
    font-size: 22px;
  }
}

<h2
  className="c0"
>
  Convert
</h2>
`;
